File: src/top2/meddelande.py
import dataclasses

from schemagen import jsontype

from top2.rolltilldelning import Rolltilldelning
from top2.roll import Roll
from top2.organisationsdel import Organisationsdel
from top2.person import Person, Passerkort, Passerbehorighet
from top2.anknytningsavtal import Anknytningsavtal


@jsontype()
@dataclasses.dataclass(kw_only=True)
class Meddelande:
    """Toppobjekt med enkla och listvärda referenser till samtliga värdeobjekt. Bra grund för meddelanden,
    även t.ex. en lysande topp-Query för ett GraphQL-gränssnitt.
    """

    anknytningsperiod: Anknytningsavtal = None
    anknytningsperioder: list[Anknytningsavtal] = None
    organisationsdel: Organisationsdel = None
    organisationsdelar: list[Organisationsdel] = None
    passerkort: Passerkort = None
    passerkortslista: list[Passerkort] = None
    passerbehorighet: Passerbehorighet = None
    passerbehorigheter: list[Passerbehorighet] = None
    person: Person = None
    personer: list[Person] = None
    roll: Roll = None
    roller: list[Roll] = None
    rolltilldelning: Rolltilldelning = None
    rolltilldelningar: list[Rolltilldelning] = None

File: CSharpReference/src/Top/Model/Meddelande.cs
namespace Top.Model;

/// <summary>
/// Top-level message object with singular and list-valued references to all value objects.
/// Good foundation for messages, also e.g., a great top-Query for a GraphQL interface.
/// </summary>
public class Meddelande
{
    // Person

    /// <summary>
    /// A single person.
    /// </summary>
    public Person? Person { get; set; }

    /// <summary>
    /// Multiple persons.
    /// </summary>
    public List<Person>? Personer { get; set; }

    // Organisationsdel

    /// <summary>
    /// A single organizational unit.
    /// </summary>
    public Organisationsdel? Organisationsdel { get; set; }

    /// <summary>
    /// Multiple organizational units.
    /// </summary>
    public List<Organisationsdel>? Organisationsdelar { get; set; }

    // Roll

    /// <summary>
    /// A single role.
    /// </summary>
    public Roll? Roll { get; set; }

    /// <summary>
    /// Multiple roles.
    /// </summary>
    public List<Roll>? Roller { get; set; }

    // Rolltilldelning

    /// <summary>
    /// A single role assignment.
    /// </summary>
    public Rolltilldelning? Rolltilldelning { get; set; }

    /// <summary>
    /// Multiple role assignments.
    /// </summary>
    public List<Rolltilldelning>? Rolltilldelningar { get; set; }

    // Anknytningsavtal (called anknytningsperiod in schema for historical reasons)

    /// <summary>
    /// A single employment contract.
    /// </summary>
    public Anknytningsavtal? Anknytningsperiod { get; set; }

    /// <summary>
    /// Multiple employment contracts.
    /// </summary>
    public List<Anknytningsavtal>? Anknytningsperioder { get; set; }

    // Passerkort

    /// <summary>
    /// A single access card.
    /// </summary>
    public Passerkort? Passerkort { get; set; }

    /// <summary>
    /// Multiple access cards.
    /// </summary>
    public List<Passerkort>? Passerkortslista { get; set; }

    // Passerbehorighet

    /// <summary>
    /// A single access privilege.
    /// </summary>
    public Passerbehorighet? Passerbehorighet { get; set; }

    /// <summary>
    /// Multiple access privileges.
    /// </summary>
    public List<Passerbehorighet>? Passerbehorigheter { get; set; }
}
